swiftsimio/subset_writer.py

```diff
@@ -0,0 +1,172 @@
+    """
+    Return number of elements of the specified dataset selected by the mask
+
+    Parameters
+    ----------
+    mask : SWIFTMask
+        the mask used to define subset that is written to new snapshot
+    dataset_name : str
+        the name of the dataset we're interested in. This is the name from the
+        hdf5 file (i.e. "PartType0", rather than "gas")
+
+    Returns
+    -------
+    int
+        size of the subset of the dataset selected by the mask
+    """
+    legend = {"Cells": 0, 
+              "PartType0": mask.gas_size if hasattr(mask, "gas_size") else 0, 
+              "PartType1": mask.dark_matter_size if hasattr(mask, "dark_matter_size") else 0, 
+              "PartType2": 0, 
+              "PartType3": 0, 
+              "PartType4": mask.stars_size if hasattr(mask, "stars_size") else 0, 
+              "PartType5": mask.black_holes_size if hasattr(mask, "black_holes_size") else 0}
+    for key in legend.keys():
+        if key in dataset_name:
+            return legend[key]
+
+def get_mask_label(mask: SWIFTMask, dataset_name: str):
+    """
+    Return appropriate mask for appropriate dataset
+
+    Parameters
+    ----------
+    mask : SWIFTMask
+        the mask used to define subset that is written to new snapshot
+    dataset_name : str
+        the name of the dataset we're interested in. This is the name from the
+        hdf5 file (i.e. "PartType0", rather than "gas")
+
+    Returns
+    -------
+    np.ndarray
+        mask for the appropriate dataset
+
+    """
+    legend = {"Cells": 0, 
+              "PartType0": mask.gas if hasattr(mask, "gas") else 0, 
+              "PartType1": mask.dark_matter if hasattr(mask, "dark_matter") else 0, 
+              "PartType2": 0, 
+              "PartType3": 0, 
+              "PartType4": mask.stars if hasattr(mask, "stars") else 0, 
+              "PartType5": mask.black_holes if hasattr(mask, "black_holes") else 0}
+    for key in legend.keys():
+        if key in dataset_name:
+            return legend[key]
+
+def write_datasubset(infile, outfile, mask: SWIFTMask, dataset_names):
+    """
+    Writes subset of all datasets contained in snapshot according to specified mask
+    Parameters
+    ----------
+    infile : h5py.File
+        hdf5 file handle for input snapshot
+    outfile : h5py.File
+        hdf5 file handle for output snapshot
+    mask : SWIFTMask
+        the mask used to define subset that is written to new snapshot
+    dataset_names : list of str
+        names of datasets found in the snapshot
+    """
+    if mask is not None:
+        for name in dataset_names:
+            if "Cells" not in name:
+                # get output dtype and size 
+                first_value = infile[name][0]
+                output_type = first_value.dtype
+                output_size = first_value.size
+                mask_size = compute_mask_size(mask, name)
+                if output_size != 1:
```

Comment:
You can probably use `infile[name].ndim` here instead?

Reply:
Tried this, didn't work. I'm looking for the _size variable (like mask.gas_size) here. I've reworked this function and the one slightly below into a single one which will select the correct attribute.